ObjectCapabilities/revokableForwarder.rho

```diff
@@ -31,14 +39,67 @@ new revokFactory in {
        * Messages to be forwarded are sent here. If the forwarder
        * if still active the message is forwarded, and an ack is
        * sent. If the forwarder is inactive, no forwarding happens
-       * and no ack is sent.
+       * and no ack is sent. The ack is just the forwarder saying,
+       * Your message got here and I passed it on. It is _not_ a
+       * guarantee that your message reaches an ultimate downstream
+       * capability in the case that several forwarders are composed.
        */
       contract forward(@message, ack) = {
         // for (_ <! activeCh) {
         for (_ <- activeCh) {
           activeCh!(Nil) |
+          ack!(Nil) |
           capability!(message)
         }
+      } |
+
+      /**
+       * Overloaded forward method nearly identical to the
+       * previous, but no ack is sent.
+       */
+       contract forward(@message) = {
+         // for (_ <! activeCh) {
+         for (_ <- activeCh) {
+           activeCh!(Nil) |
+           capability!(message)
+         }
+       }
+    }
+  } |
+
+  // Demo of using the forwarder.
+  new superpower,
+    killSwitchCh,
+    forwardCh,
+    ack,
+    stdout(`rho:io:stdout`)
+    in {
+
+    // Silly superpower that we'll wrap with a forwarder
+    // Explicit tupling is not ideal. But I don't see a better option.
+    contract superpower(@(payload /\ String, return)) = {
+      @return!("Superpower Invoked with " ++ payload)
+    } |
+
+    // Create a forwarder for the superpower
+    // revokFactory!(*superpower, await *killSwitchCh, await *forwardCh) {
+    revokFactory!(*superpower, *killSwitchCh, *forwardCh) |
+    for (killSwitch <- killSwitchCh; forward <- forwardCh) {
+
+      // Call the forwarder
+      // Notice the explicit tupling of the entire message
+      // including return channel that goes on to the superpower
+      forward!(("Apples", *stdout), *ack) |
+      for (_ <- ack) {
+
+        // Revoke the forwarder
+        killSwitch!(Nil, *ack) |
+        for (_ <- ack) {
```

Comment:
This sort of hang / silent failure is a real PITA in development.

In E renditions of the caretaker (aka revokable forewarder), calls after the kill switch result in a throw rather than a hang. The possibility of a throw is part of the uniform calling convention in E. I'm not sure what's a good uniform calling convention for rholang. I've been using `reject, resolve` return channels lately.


ref section 4.3 of [Paradigm Regained: Abstraction Mechanisms for Access Control][1] by Mark S. Miller, Jonathan S. Shapiro

[1]: http://www.erights.org/talks/asian03/

Reply:
It seems like we've swapped sides since https://github.com/JoshOrndorff/librho/issues/7#issuecomment-432496586

I think I like `reject`, `resolve`